File: 438.py
# 438. 找到字符串中所有字母异位词
# https://leetcode-cn.com/problems/find-all-anagrams-in-a-string/
from typing import List
import collections
# 思路：滑动窗口 维护一个数量字典 每一滑动 移动改变计数器
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        #滑动窗口
        m = len(p)
        n = len(s)
        ans = []
        if m > n:
            return ans
        
        dp = 0

        time_1 = {i:0 for i in p}
        time_2 = time_1.copy()
        for i in p:
            time_1[i] += 1
        
        
        for j in range(m):
            tmp = time_2.get(s[j], -1)
            if tmp >= 0:
                time_2[s[j]] += 1
                if tmp < time_1[s[j]]:
                    dp += 1

        if dp == m:
            ans.append(0)
        for j in range(1, n - m + 1):
            tmp = time_2.get(s[m + j - 1], -1)
            if tmp >= 0:
                time_2[s[m + j - 1]] += 1
                if tmp < time_1[s[m + j - 1]]:
                    dp += 1
            
            tmp = time_2.get(s[j - 1], -1)
            if tmp >= 0:
                time_2[s[j - 1]] -= 1
                if tmp <= time_1[s[j - 1]]:
                    dp -= 1
            
            if dp == m:
                ans.append(j)
        
        return ans
        
        

File: leetcode_c_plus_plus/438.cpp
// 438. 找到字符串中所有字母异位词
// https://leetcode-cn.com/problems/find-all-anagrams-in-a-string/
#include<map>
#include<set>
#include<vector>
#include<cmath>
#include<numeric>
using namespace std;
#define INT_MAX 2147483637
#define INT_MIN (-INT_MAX - 1)
/**
@File : 438.cpp
@Time : 2021/11/28 20:59:54
@Author : YuMin Zhang
@State : Indepeedent | Depedent | Half-Depedent
@Thinking :
**/

class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        int sLen = s.size();
        int pLen = p.size();
        vector<int> ans;
        int count[26] = {0};
        if(pLen > sLen) return ans;
        
        for(int j = 0; j < pLen; j++){
            count[p[j] - 'a']++;
            count[s[j] - 'a']--;
        }
        int differ = 0;
        for(int i = 0; i < 26; i++){
            if(count[i] != 0) differ++;
        }
        if(differ == 0) ans.push_back(0);
        for(int i = 0; i < sLen - pLen; i++){
            if(count[s[i] - 'a'] == -1) differ--;
            else if(count[s[i] - 'a'] == 0) differ++;
            count[s[i] - 'a']++;
            if(count[s[i + pLen] - 'a'] == 1) differ--;
            else if(count[s[i + pLen] - 'a'] == 0) differ++;
            count[s[i + pLen] - 'a']--;
            if(differ == 0) ans.push_back(i + 1); 
        }
        return ans;

    }
};